service/storage_service.cc

```diff
@@ -539,6 +539,9 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
                 update_topology(host_id, ip, rs);
                 tmptr->add_replacing_endpoint(replaced_host_id, host_id);
                 co_await update_topology_change_info(tmptr, ::format("replacing {}/{} by {}/{}", replaced_id, *existing_ip, id, ip));
+                if (existing_ip == ip && !_gossiper.is_alive(*existing_ip)) {
+                    co_await _gossiper.real_mark_alive(*existing_ip);
```

Comment:
Is this atomic with token ring change?

IIUC yes because we call `sync_raft_topology_nodes` under token metadata lock always

Reply:
There's a lot of locking going on here, and I'm worried that we won't enter some deadlock
because
we lock token_metadata
then we lock endpoint inside real_mark_alive
this calls on_alive subscriptions -- are we sure one of these subscriptions won't try to lock something again?

Reply:
* `on_alive` handlers are called under endpoint lock acquired inside `real_mark_alive` itself. This shouldn't deadlock with gossiper locks since `sync_raft_topology_nodes` is never called with gossiper locks acquired.
* in raft topology mode we shouldn't be touching `token_metadata` in gossiper event handlers, so shouldn't deadlock with `token_metadata_lock` either.

Reply:
> Is this atomic with token ring change?

Not actually, we're marking the node alive before we've published a new `token_metadata` version. Is it really important?

Reply:
> Not actually, we're marking the node alive before we've published a new token_metadata version. Is it really important?

https://github.com/scylladb/scylladb/issues/17421#issuecomment-2022433343
Basically reintroducing the "On gossiper path:" case. Which may give inconsistency.

Reply:
This is no worse than it was with gossiper. Not clear how to fix -- the node can become alive anytime on its own since it may already advertised itself in `join_node_response_handler`.